packages/gatsby/src/utils/start-server.ts

```diff
@@ -310,7 +348,46 @@ module.exports = {
 
       if (page) {
         try {
+          let serverDataPromise: Promise<IServerData> = Promise.resolve({})
+
+          if (page.mode === `SSR`) {
+            // get dynamic serverModule$
+            const stats = await waitForServerCompilation()
+
+            if (
+              !stats ||
+              !stats.compilation.entrypoints.has(page.componentChunkName)
+            ) {
+              report.error(
+                `Error loading a result for the page query in "${requestedPagePath}" / "${potentialPagePath}". getServerData threw an error.`
+              )
+            }
+
+            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+            const componentHash = stats!.compilation.entrypoints
+              .get(page.componentChunkName)!
+              .getEntrypointChunk().hash as string
+            const modulePath = path.resolve(
+              `${program.directory}/${ROUTES_DIRECTORY}/${page.componentChunkName}.js`
+            )
+
+            // if webpack compilation is diff we delete old cache
+            if (
+              previousHashes.has(page.componentChunkName) &&
+              previousHashes.get(page.componentChunkName) === componentHash
+            ) {
+              delete require.cache[modulePath]
+            }
+
+            const mod = require(modulePath)
+
+            previousHashes.set(page.componentChunkName, componentHash)
+
+            serverDataPromise = getServerData(req, page, potentialPagePath, mod)
+          }
+
           let pageData: IPageDataWithQueryResult
+          // TODO move to query-engine
```

Comment:
not sure if we want to move stuff from develop to use engines (because first we would need to generate those)

Reply:
in the long run we should :) but not right now